src/mscorlib/src/System/TimeZoneInfo.cs

```diff
@@ -1341,28 +1370,135 @@ private TimeZoneInfo(byte[] data, string id, Boolean dstDisabled)
             // time span {4/2/2006@00:00:00.00000 - 10/28/2006@23:59:59.99999}
-            DateTime date = dateTime.Date;
+            DateTime date;
+            if (dateTimeisUtc)
+            {
+                date = (dateTime + BaseUtcOffset).Date;
+            }
+            else
+            {
+                date = dateTime.Date;
+            }
 
             for (int i = 0; i < m_adjustmentRules.Length; i++) {
-                if (m_adjustmentRules[i].DateStart <= date && m_adjustmentRules[i].DateEnd >= date) {
-                    return m_adjustmentRules[i];
+                AdjustmentRule rule = m_adjustmentRules[i];
+                AdjustmentRule previousRule = i > 0 ? m_adjustmentRules[i - 1] : rule;
+                if (IsAdjustmentRuleValid(rule, previousRule, dateTime, date, dateTimeisUtc)) { 
+                    return rule;
                 }
             }
 
             return null;
         }
 
+        /// <summary>
+        /// Determines if 'rule' is the correct AdjustmentRule for the given dateTime.
+        /// </summary>
+        private bool IsAdjustmentRuleValid(AdjustmentRule rule, AdjustmentRule previousRule,
+            DateTime dateTime, DateTime dateOnly, bool dateTimeisUtc)
+        {
+            bool isAfterStart;
+            if (rule.DateStart.Kind == DateTimeKind.Utc)
+            {
+                DateTime dateTimeToCompare;
+                if (dateTimeisUtc)
+                {
+                    dateTimeToCompare = dateTime;
+                }
+                else
+                {
+                    dateTimeToCompare = ConvertToUtc(dateTime,
+                        // use the previous rule to compute the dateTimeToCompare, since the time daylight savings "switches"
+                        // is based on the previous rule's offset
+                        previousRule.DaylightDelta, previousRule.BaseUtcOffsetDelta);
+                }
+
+                isAfterStart = dateTimeToCompare >= rule.DateStart;
+            }
+            else
+            {
+                // if the rule's DateStart is Unspecified, then use the whole-date portion
+                isAfterStart = dateTime >= rule.DateStart;
+            }
+
+            bool isBeforeEnd;
+            if (rule.DateEnd.Kind == DateTimeKind.Utc)
+            {
+                DateTime dateTimeToCompare;
+                if (dateTimeisUtc)
+                {
+                    dateTimeToCompare = dateTime;
+                }
+                else
+                {
+                    dateTimeToCompare = ConvertToUtc(dateTime, rule.DaylightDelta, rule.BaseUtcOffsetDelta);
+                }
+
+                isBeforeEnd = dateTimeToCompare <= rule.DateEnd;
+            }
+            else
+            {
+                // if the rule's DateEnd is Unspecified, then use the whole-date portion
+                isBeforeEnd = dateTime <= rule.DateEnd;
+            }
+
+            return isAfterStart && isBeforeEnd;
```

Comment:
How about adding a check for ```isAfterStart``` before the whole ```isBeforeEnd``` section so that we don't have to do any of the latter section if isAfterStart is false? e.g.
```C#
if (!isAfterStart)
    return false;

bool isBeforeEnd;
...
return isBeforeEnd;
```